assimilation_code/modules/assimilation/assim_tools_mod.pf.f90

```diff
@@ -1447,109 +1719,131 @@ subroutine filter_assim(ens_handle, obs_ens_handle, obs_seq, keys,           &
    OBS_UPDATE: do j = 1, num_close_obs
       obs_index = close_obs_ind(j)
 
-      if (my_obs_indx(obs_index) <= i) cycle OBS_UPDATE
-
-      ! Only have to update obs that have not yet been used
-      if(my_obs_indx(obs_index) > i) then
-
-         ! If the forward observation operator failed, no need to
-         ! update the unassimilated observations
-         if (any(obs_ens_handle%copies(1:ens_size, obs_index) == MISSING_R8)) cycle OBS_UPDATE
+      ! Skip if regularization reaches threashold value 
+      if (filter_kind == 9) then
+         if ( beta_y(obs_index) == beta_max ) cycle OBS_UPDATE
+      end if
 
-         ! Compute the distance and the covar_factor
-         cov_factor = comp_cov_factor(close_obs_dist(j), cutoff_rev, &
-            base_obs_loc, base_obs_type, my_obs_loc(obs_index), my_obs_kind(obs_index))
+      ! If the forward observation operator failed, no need to 
+      ! update the unassimilated observations 
+      if (any(obs_ens_handle%copies(1:ens_size, obs_index) == MISSING_R8)) cycle OBS_UPDATE
 
-         ! if external impact factors supplied, factor them in here
-         ! FIXME: this would execute faster for 0.0 impact factors if
-         ! we check for that before calling comp_cov_factor.  but it makes
-         ! the logic more complicated - this is simpler if we do it after.
-         if (adjust_obs_impact) then
-            impact_factor = obs_impact_table(base_obs_type, my_obs_kind(obs_index))
-            cov_factor = cov_factor * impact_factor
-         endif
+      ! Compute the distance and the covar_factor
+      cov_factor = comp_cov_factor(close_obs_dist(j), cutoff_rev, &
+         base_obs_loc, base_obs_type, my_obs_loc(obs_index), my_obs_kind(obs_index))
 
-         if(cov_factor <= 0.0_r8) cycle OBS_UPDATE
+      ! SEEMS TO BE NEEDED FOR VERTICAL LOCALIZATION IN BGRID
+      if (filter_kind == 9) then
+         if ( cov_factor == 0.0_r8 ) cycle OBS_UPDATE
```

Comment:
same comment - filter_kinds other than 9 need to cycle if cov_factor is 0, also it should be moved so impact_factor is applied first before testing.